### PR TITLE
[RFC][docker] add toolchain layer

### PR DESCRIPTION
**Pull request by sausagee**
## Motivation
There are multiple failure points in our CI/CD pipeline. Building the
prod images requires network access to a plephora of sources. Some may
be necessary to get the latest source code, but some, such as 3rd party
tooling binaries, could be and should be cached. This will reduce our
exposure to external dependencies.

The end goal is to build the prod images from the cache layer without
network. For the time being, we want to be able to cache as many
artifacts as possible to limit exposure.

This toolchain cache layer (toolchain_rust) does the following:

    install the matching Rust compiler outlined in libra/rust-toolchain
    prefetch crates
    install 3rd party binaries, such as kubectl

The layer tag includes Rust compiler version and first 8 digits of
sha256 sum of Cargos.lock. For example, at commit 1a8c6c564 in
libra/libra, the tag is 1.42.0-0151b8a7

## Test Plan
- CI
- Built all images locally
- Local CTI

### CODE REVIEW

**Comment by sausagee** [comment]
This is the first iteration with a proof-of-concept in https://github.com/calibra/libra-ops/pull/525.  

I'd love to gather feedbacks, particularly the caching aspect, before expanding it further (like Cluster Test images).

File: docker/toolchain/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+WORKDIR /staging/libra
+COPY . /staging/libra
+COPY rust-toolchain /staging/rust-toolchain
+RUN rustup install $(cat rust-toolchain)
+
+# Prefetch crates
+RUN cargo fetch
+RUN set -e; shasum -a 256 Cargo.lock | cut -d' ' -f1 > /staging/crates_cache.sha256
+WORKDIR /staging
+RUN rm -rf /staging/libra
```
Comment (rexhoffman) [inline docker/toolchain/Dockerfile:34]: perhaps add a testing layer and push this layer as soon as it's built.

**Comment by mgorven** [comment]
So we can avoid the separate toolchain image with the `--target` build parameter. The Dockerfiles would go something like this: https://github.com/mgorven/libra/commit/708884450a2dfad0cb7b088f654c59315c14002d

And then CircleCI would run:

    docker build --target crate_cache --tag libra_crates -f libra/docker/validator/Dockerfile libra

It would then tag `libra_crates:latest` with the tool chain and Cargo.lock SHA somehow and push to ECR. The rest of the build would then stay the same.

**Comment by mgorven** [comment]
Um, although this will cache miss with any change in the libra repo, which I suppose defeats the point. I think the real solution here is vendoring crates in our infra.

**Pull request closed without merging**